README.md
# The App Scrolls for creating and transforming Rails apps

```
             ___               ____             ____  
            / _ | ___  ___    / __/__________  / / /__
           / __ |/ _ \/ _ \  _\ \/ __/ __/ _ \/ / (_-<
          /_/ |_/ .__/ .__/ /___/\__/_/  \___/_/_/___/
               /_/  /_/                               
```

The App Scrolls is a magical tool to generate new Rails and modify existing Rails applications (coming) to include your favourite, powerful magic. Authentication, testing, persistence, javascript, css, deployment, and templating - there's a magical scroll for you.

* Follow on twitter [@appscrolls][9]
* [![Build Status](https://secure.travis-ci.org/drnic/appscrolls.png?branch=master)](http://travis-ci.org/drnic/appscrolls)
* [![Dependency Status](https://gemnasium.com/drnic/appscrolls.png?branch=master)](https://gemnasium.com/drnic/appscrolls)

An example application that was built by the App Scrolls is at [https://github.com/drnic/mydemoapp][14]. The generated README shows all the scrolls that were included.

## Installation

Installation is simple:

    gem install appscrolls

## Usage

The primary usage of the `appscrolls` gem is to utilize its interactive terminal command to build a new Rails application. To get started, you can simply run the command thusly:

    appscrolls new APP_NAME
    scrolls new APP_NAME

Where `APP_NAME` is the directory in which you wish to create the app (it mirrors the Rails creation syntax). You will then be guided through the scroll selection process and subsequently the Rails app generator will automatically run with the template and all appropriate command line options included.

To transform an existing Rails app, you ... wait, that's not implemented yet. But since the "apply template" feature of `rails new APP_NAME -m template.rb` is implemented in Thor, I mean, how hard could it be?*

### Available Scrolls

The current available scrolls grouped by category:

* administration: active_admin, rails_admin
* assets: jquery, prototype
* deployment: eycloud, eycloud_recipes_on_deploy, git, github, passenger, thin, unicorn
* persistence: mysql, postgresql, redis, sqlite3
* stylesheet: twitter_bootstrap
* templating: simple_form
* testing: capybara, cucumber, rspec, test_unit
* worker: delayed_job, resque
* other: env_yaml, guard, rails_basics, split

### Specifying Scrolls

If you wish to skip the interactive scroll selector, you may provide instead a list of scrolls with the `-s` or `--scrolls` option:

    scrolls new APP_NAME -s twitter_bootstrap mysql resque
    scrolls new APP_NAME --scrolls postgresql github eycloud

This will automatically generate a Rails template with the provided scrolls and begin the app generator.

### Listing Scrolls

You can also print out a simple list of scrolls:

    scrolls list

Or print out a list of scrolls for a specific category:

    scrolls list persistence

### User Scrolls

You can define an environment variable APPSCROLLS_DIR with a path to your local app scrolls:

    export APPSCROLLS_DIR=~/.scrolls

### Scroll Configuration

If you wish to skip any configuration questions, you can provide the path to a configuration file with the `-c` or `--config` option:

    scrolls new APP_NAME -c ~/.scrolls_config.rb
    
This config file is a script that can provide defaults for scrolls by setting config.SCROLL_NAME.OPTION_NAME.  For example:
```
config.postgresql.pg_username = 'root'
config.postgresql.pg_password = ''
config.guard.guard_notifications = false
```

Or instead of passing a flag each time, just set the APPSCROLLS_CONFIG environment variable to include it every time:
    export APPSCROLLS_CONFIG=~/.scrolls_config.rb

## Deployment Support

Web applications are boring if they aren't running proudly on the internet. The App Scrolls make this automatic for your favourite providers!

### Engine Yard

Scroll: `eycloud`

Example:

```
scrolls new mydemoapp -s twitter_bootstrap unicorn postgresql resque github eycloud
```

The created application above will be automatically stored in a git private/public repository on GitHub and then deployed to [Engine Yard Cloud][6].

When deploying to [Engine Yard Cloud][6] you will need to choose:

* `unicorn` or `passenger` for your app server
* `mysql` or `postgresql` for your SQL DB

When choosing from the following scrolls, your Engine Yard Cloud environment will be automatically upgraded/configured with Chef recipes. 

* `resque` - add utility instances called `resque` for workers [see [readme][15] for more information]
* `redis` - add a utility called `redis` to have a dedicated redis DB; else it is run on your DB master or Solo instance
* `delayed_job` - add utility instances called `delayed_job` or `dj` for workers [see [readme][16] for more information]

Note: Resque is recommended instead of Delayed Job. Soon, Sidekiq will be recommended over both, and 

Please open an Issue if you want an alternate option (`puma` or `thin`; or `mongodb`)

### Heroku

The App Scrolls needs a Heroku Master to support Heroku for the App Scrolls. 

There is some initial work in the [current scrolls][11] and the [archived/unsupported scrolls][12]

### CloudFoundry

The App Scrolls needs a CloudFoundry Master to support CloudFoundry for the App Scrolls. 

## Authoring Scrolls of Magical Mystery

Create new scrolls using:

    thor :new scroll-name

Submitting a scroll is actually a very straightforward process. Scrolls are made of up **template code** and **YAML back-matter** stored in a ruby file. The `__END__` parsing convention is used so that each scroll is actually a valid, parseable Ruby file. The structure of a scroll looks something like this:

```ruby
gem 'supergem'

after_bundler do
  generate "supergem:install"
end

__END__

category: templating
name: SuperGem
description: Installs SuperGem which is useful for things
author: mbleigh
```

It's really that simple. The gem has RSpec tests that automatically validate each scroll in the repository, so you should run `rake spec` as a basic sanity check before submitting a pull request. Note that these don't verify that your scroll code itself works, just that App Scrolls could properly parse and understand your scroll file.

## History

This project is an old fashioned fork of [Michael Bleigh][5]'s [Rails Wizard][4]. A new name, new project, and new purpose. 

This project wouldn't exist without Michael having created [Rails Wizard][4] during Rails Rumble and maintaining and upgrading it for a long time. Sadly support dropped off, several recipes did not work with Rails 3.1+, 

[Dr Nic][7] originally worked on [Rails Wizard][4] to provide [Engine Yard Cloud][6] support, his employer and his favourite hosting platform. He also merged in a lot of recipes from other forks, and added new recipes for modern projects.

Support for Engine Yard Cloud meant integration with Chef Recipes. This meant confusing language - Rails Wizard Recipes and Chef Recipes. He decided that wizards don't use recipes - they use scrolls. Alchemists use recipes. And screw alchemists and their dinky potions. Recipes became Scrolls.

## Future

* Automatically setup Continuous Integration for new applications - branches "jenkins"
* Interactive mode is a wizard by categories "pick A, B, C or none"
* Apply scrolls to existing Rails applications - branch "[apply_scrolls][13]"*
* Scrolls work or fail fast on Heroku
* Scrolls work or fail fast on CloudFoundry
* Scrolls generate their own README - branch "readmes"
* 3rd party services/add-ons enabled within deployment platform or directly with service
* Padrino / Sinatra applications
* Non-Ruby applications (Lithium for PHP, etc)

Missing scrolls

* MongoDB - branch "mongodb"
* OmniAuth - branch "omniauth"
* Sidekiq - branch "sidekiq"

How hard could it be?

* `*` 'How hard could it be to transform applications?' - pretty hard. Scrolls need to be aware of the current code base, rather than merely the list of other scrolls being used to create a new app. Scrolls also need to know about versions of Rails rather than just latest rails.

## Thanks

ASCII banner - http://www.network-science.de/ascii/ using 'smslant' font.

## License

App Scrolls and its scrolls are distributed under the MIT License. See [MIT_LICENSE][10] for the actual words.

[1]:http://appscrolls.org/
[2]:https://github.com/drnic/appscrolls
[2]:https://github.com/drnic/appscrolls/tree/master/scrolls
[4]:https://github.com/intridea/rails_wizard
[5]:https://github.com/mbleigh
[6]:http://www.engineyard.com/products/cloud
[7]:http://drnicwilliams.com
[9]:https://twitter.com/appscrolls
[10]:https://github.com/drnic/appscrolls/blob/master/MIT_LICENSE
[11]:https://github.com/drnic/appscrolls/tree/master/scrolls
[12]:https://github.com/drnic/appscrolls/tree/master/scrolls/zzz
[13]:https://github.com/drnic/appscrolls/tree/apply_scrolls
[14]:https://github.com/drnic/mydemoapp
[15]:https://github.com/engineyard/eycloud-recipe-resque#readme
[16]:https://github.com/engineyard/eycloud-recipe-delayed_job#readme


ChangeLog.md
# ChangeLog

## v0.11

* `sidekiq` - adds sidekiq gem & web dashboard (hidden via secret string); can deploy worker to Cloud Foundry

### v0.11.1

* `sidekiq` - correctly create the app/workers directory

### v0.11.2

* `redis` - use integer for redis database number [thx @mperham]

## v0.10

* `puma` - added basic support for Puma; also Cloud Foundry will use Puma automatically

### v0.10.1

* `cfoundry` renamed to `cf`
* `cf` - automatically creates & binds postgresql/mysql if selected; deletes existing app of same name
* installation forces installation of latest bundler 1.3.4+ & rails 3.2.13+

## v0.9

* `Cloud Foundry` - added support for uploading to Cloud Foundry
* `github` - now uses `hub` project instead of `github` gem
* `postgresql` - database.yml includes "host: localhost" for each env

## v0.8

* `postgresql` - supported on `eycloud` (Currently PostgreSQL 9.1)
* `eycloud` - can boot clusters, in addition to solos
* `mysql` - always generates `sample.rake`
* bin/scrolls is short-form of bin/appscrolls
* `puma` & `thin` - moved into zzz archive until Puma is released on EY Cloud; or another supporting platform added
* `spork` - moved into zzz archive as part of simplification

### v0.8.1

* `twitter_bootstrap` - includes `rails_basics`
* `eycloud` - only using `ey_cli` (`engineyard` gem has `net-ssh` conflict with `chef` gem)
* Display scrolls that will be used - explicitly requested (green) and implicitly included (yellow)
* `eycloud` - only unicorn and passenger initially supported
* `eycloud` - don't check for available app name - this should be implemented in ey_cli

### v0.8.2

PUBLIC RELEASE! http://drnicwilliams.com/2012/04/10/instant-new-rails-applications-with-the-app-scrolls/

* `delayed_job` - always get admin dashboard
* `mysql` - always create local databases

### v0.8.3

* `resque` - Use "resque_admin_secret" instead of hardcoded string

### v0.8.4

* `github` - if private, then create private repo [thx @CraigCottingham]
* `thin` - moved to active
* `rails_admin` - new scroll [thx @sferik]
* `twitter_bootstrap` - Added `therubyracer` gem as a dependency [thx @jackdempsey]
* added support for scrolls with dashes in their names [thx @grk]

## v0.7

New name! "App Scrolls"

* `split` - AB testing promoed in RailsCasts
* `spork` - clean running of tests
* `resque` - always includes admin dashboard

Fixed

* `cucumber` - creates databases before running installer

Internal

* Replace ZenTest with Guard
* Add 'rake list:categories' task

### v0.7.1

* Force rails ~> 3.2.2 as a dependency

## v0.6

* `twitter_bootstrap` - automatically includes `simple_form`; includes flash msg
* `github` - asks for a new repo name if cannot create a repository
* `rails_basics` - default flash message
* `jquery` & `prototype` - former is already default in Rails 3.1+ now

New
* `guard` - guard support for all supported scrolls
* `postgresql`
* All untested scrolls moved into scrolls/zzz

## v0.5

* `twitter_bootstrap` - based on public RailsCast
* `eycloud` - now using `ey_cli` to create/boot environments

## v0.4

* SCROLLS ARE IN, "recipes" are out. Wizards use scrolls. Alchemists use recipes.
* Scrolls have no dependencies on `eycloud` or `eycloud_recipes_on_deploy`; these are optional
* `github` - GitHub repository creation
* `resque` + `delayed_job` can install their admin consoles
* `sidekiq` - high performance, low cost alternative to `resque` or `delayed_job`

## v0.3

* Engine Yard support - Resque
* Removed a bunch of recipes that didn't see useful/common

### v0.3.1

* Resque recipe works a treat.

Try this on EY Cloud! `appscrolls new demomysql -r sqlite3 mysql resque rails_basics git`

## v0.2

Merged lots of branches/forks

## v0.2.1

* Cleaned up many recipes
* Added :website attribute to recipes
* Convert all recipes to use supported categories



scrolls/cf.rb
# Scroll is based on the following documentation
# * http://blog.cloudfoundry.com/2012/03/15/using-cloud-foundry-services-with-ruby-part-2-run-time-support-for-ruby-applications/
# * http://blog.cloudfoundry.com/2012/04/19/deploying-jruby-on-rails-applications-on-cloud-foundry/

require "yaml"

@vmc_version = '0.3.23'
@name = File.basename(File.expand_path("."))
@cf_ruby_runtime = "ruby19" # might change later in scrolls

gem 'vmc', "~> #{@vmc_version}"
gem  'cf-runtime'

known_services = %w[postgresql mysql redis mongodb]
required_dbs = %w[mysql postgresql]

selected_db = required_dbs.find { |db| scroll? db }
unless selected_db
  say_custom "cf", "Please include a DB choice from: #{required_dbs.join ", "}"
  exit_now = true
end

$cf_manifest = {"applications"=>
  {"."=>
    {"name"=>@name,
     "framework"=>
      {"name"=>"rails3",
       "info"=>
        {"mem"=>"256M", "description"=>"Rails Application", "exec"=>nil}},
     "url"=>"${name}.${target-base}",
     "runtime"=>@cf_ruby_runtime,
     "mem"=>"256M",
     "instances"=>1,
     "services"=>{}}}}

known_services.each do |service|
  if scroll? service
    $cf_manifest["applications"]["."]["services"]["#{@name}-#{service}"] = {"type"=>service}
  end
end

db_username = config['pg_username'] || 'root'
db_password = config['pg_password'] || ''

exit 1 if exit_now

after_bundler do
  cf_delete_app @name

  # Desirable to vendor everything
  run "bundle package"

  production = <<-YAML.gsub(/^\s{2}/, '')
  production:
    adapter: #{selected_db}
    <% db_svc = CFRuntime::CloudApp.service_props('#{selected_db}') %>
    database: <%= db_svc[:database] rescue '#{project_name}_production' %>
    username: <%= db_svc[:username] rescue '#{db_username}' %>
    password: <%= db_svc[:password] rescue '#{db_password}' %>
    <%= db_svc && db_svc[:host] ? "host: \#{db_svc[:host]}" : "" %>
    <%= db_svc && db_svc[:port] ? "port: \#{db_svc[:port]}" : "" %>
  YAML
  append_file "config/database.yml", production
end

after_everything do
  create_file "manifest.yml", $cf_manifest.to_yaml

  run "rake assets:precompile"
  if jruby?
   run "warble"
    run "mkdir -p deploy"
    run "cp #{project_name}.war deploy/"
  end
  run_vmc "push #{project_name} --runtime #{@cf_ruby_runtime} --path . --no-start"
  run_vmc "env-add #{project_name} BUNDLE_WITHOUT=assets:test:development"
  run_vmc "start #{project_name}"
end

def vmc
  "vmc _#{@vmc_version}_"
end

def run_vmc(command)
  run "#{vmc} #{command}"
end

def cf_delete_app(name)
  run %Q{#{vmc} apps | grep " #{name} " && #{vmc} delete #{name}}
end

def cf_standalone_command(key, name, command, services={})
  cf_manifest = {"applications"=>
    {"."=>
      {"name"=>name,
       "framework"=>
        {"name"=>"standalone",
         "info"=>
          {"mem"=>"64M", "description"=>"Standalone Application", "exec"=>nil}},
       "url"=>nil,
       "runtime"=>@cf_ruby_runtime,
       "command"=>command,
       "mem"=>"256M",
       "instances"=>1,
       "services"=>services}}}
  create_file "manifest.#{key}.yml", cf_manifest.to_yaml
  "manifest.#{key}.yml"
end

__END__

name: Cloud Foundry
description: Prepare codebase and perform initial deploy to a Cloud Foundry target
website: http://cloudfoundry.org
author: drnic

requires: []
run_after: [postgresql, mysql]
run_before: []

category: deployment
